pubtools/_pulp/services/pulp.py

```diff
@@ -0,0 +1,69 @@
+import threading
+import os
+import logging
+import warnings
+
+from pubtools import pulplib
+
+from .base import Service
+
+LOG = logging.getLogger("pubtools.pulp")
+
+
+class PulpClientService(Service):
+    """A service providing a Pulp client.
+
+    If this service is inherited, Pulp-related arguments become mandatory
+    in order to run the task.
+    """
+
+    def __init__(self, *args, **kwargs):
+        self.__lock = threading.Lock()
+        self.__instance = None
+        super(PulpClientService, self).__init__(*args, **kwargs)
+
+    def add_service_args(self, parser):
+        super(PulpClientService, self).add_service_args(parser)
+
+        group = parser.add_argument_group("Pulp environment")
+        group.add_argument("--pulp-url", help="Pulp server URL", required=True)
+        group.add_argument("--pulp-user", help="Pulp username", default=None)
+        group.add_argument(
+            "--pulp-password",
+            help="Pulp password (or set PULP_PASSWORD environment variable)",
+            default=None,
+        )
+        group.add_argument(
+            "--pulp-insecure",
+            action="store_true",
+            help="Allow unverified HTTPS connection to Pulp",
+        )
+
+    @property
+    def pulp_client(self):
+        """A Pulp client used during task, instantiated on demand."""
+        with self.__lock:
+            if not self.__instance:
+                self.__instance = self.__get_instance()
```

Comment:
I don't understand this. 

looks like __get_instance would only be called when there is no __instance bound, but the clearRepo task inherits 4 services and it gets all kinds of clients, I don't know why:
1. isn't __get_instance overwritten by the last parent? 
2. __get_instance called once and it assigns the instance to self._instance and it shouldn't be called again, why it get all 4 service clients?

Reply:
I think what you're missing is the special behavior of `__` prefix:

In Python, if you name a class property or method with leading `__`, it's mangled to include the class name.  That's specifically so that you can reuse the same name in multiple classes within a hierarchy, without clashes.

See this demo:

```
$ python3
Python 3.7.4 (default, Jul  9 2019, 16:32:37) 
[GCC 9.1.1 20190503 (Red Hat 9.1.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> class Klass:
...   __my_variable = 'abc'
... 

# In dir(), we can see the real name used to store it is _Klass__my_variable
>>> dir(Klass)
['_Klass__my_variable', '__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__']

# From outside the class, I can't access it with the declared name
>>> Klass.__my_variable
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: type object 'Klass' has no attribute '__my_variable'

# Because it's actually named this
>>> Klass._Klass__my_variable
'abc'
```

It's good practice to use this on classes designed to be inherited, because otherwise everyone inheriting your class has to know what names you've used, so they can avoid clashes.

Reply:
ah, yeah, I completely forget this feature of python. 
I just remember prefix with double-underscore is the recommended way to make objects private in python but nobody uses it...
Never expected it could achieve such things.